### PR TITLE
blockstore: close AllKeysChan channel when done

### PR DESCRIPTION
**Pull request by mvdan**
(see commit message)

### CODE REVIEW

**Comment by mvdan** [comment]
cc @dirkmc, I can't request a review since you're not in this repo. Maybe ask to be added to the ipld org?

File: v2/blockstore/readonly.go [hunk 1]
```diff
@@ -165,7 +165,7 @@ func (b *ReadOnly) AllKeysChan(ctx context.Context) (<-chan cid.Cid, error) {
 
 	ch := make(chan cid.Cid, 5)
 	go func() {
-		done := ctx.Done()
+		defer close(ch)
```
Comment (dirkmc) [inline v2/blockstore/readonly.go:168]: I'd suggest writing out a log line when there's an error (lines 174 and 178)

File: v2/blockstore/readonly.go [hunk 2]
```diff
@@ -165,7 +165,7 @@ func (b *ReadOnly) AllKeysChan(ctx context.Context) (<-chan cid.Cid, error) {
 
 	ch := make(chan cid.Cid, 5)
```
Comment (dirkmc) [inline v2/blockstore/readonly.go:166]: Where does the number 5 come from?

**Review by dirkmc — APPROVED** [review]
LGTM 👍 

I have a couple of non-blocking suggestions in unrelated parts of the code, but they are unrelated to this PR so may make sense to address them in a follow-up PR

File: v2/blockstore/readonly.go (hunk 2 above)
Comment (mvdan) [reply v2/blockstore/readonly.go:166]: I wondered the same; this originates from @willscott's carbs at https://github.com/willscott/carbs/blob/1036c89e1fd753bc375d860725c9e29ab85b72ac/carbs.go#L161.

My guess is he wanted the channel to have some buffering for the sake of performance. I imagine that, unless measured, we should stick to a buffer size of 1, or no buffering. Any other buffer size should be carefully selected and documented, in my opinion :)

File: v2/blockstore/readonly.go (hunk 1 above)
Comment (mvdan) [reply v2/blockstore/readonly.go:168]: This is a library though, so logging requires some form of API to allow being configured - and to probably be off by default. Is there a standard across PL? I know https://github.com/ipfs/go-log, but that feels too heavy to pull into relatively low-level IPLD libraries.

Maybe I just expose an option like:

```
WithLogging(func(format string, args ...interface{})) Option
```

File: v2/blockstore/readonly.go (hunk 2 above)
Comment (ribasushi) [reply v2/blockstore/readonly.go:166]: Obligatory thread: https://twitter.com/erentantekin/status/1330957563496525824

File: v2/blockstore/readonly.go (hunk 1 above)
Comment (dirkmc) [reply v2/blockstore/readonly.go:168]: I guess really the error should be returned to the caller somehow but I assume it's not possible to change the AllKeysChan API at this stage.
Suggest we merge this PR as a strict improvement and discuss options in an issue.

File: v2/blockstore/readonly.go (hunk 2 above)
Comment (dirkmc) [reply v2/blockstore/readonly.go:166]: Agreed 👍

File: v2/blockstore/readonly.go (hunk 1 above)
Comment (mvdan) [reply v2/blockstore/readonly.go:168]: This is the blockstore interface, so we indeed can't change the method signature :) I'll raise an issue about logging.

File: v2/blockstore/readonly.go (hunk 2 above)
Comment (mvdan) [reply v2/blockstore/readonly.go:166]: I've added a TODO about this.

File: v2/blockstore/readonly.go (hunk 1 above)
Comment (mvdan) [reply v2/blockstore/readonly.go:168]: https://github.com/ipld/go-car/issues/112

**Pull request merged**